src/node.rs

```diff
@@ -179,105 +199,191 @@ type BoxedNode<K, V> = NonNull<LeafNode<K, V>>;
-        new_node.edges[0].write(child.node);
-        unsafe { NodeRef::from_new_internal(new_node, child.height + 1) }
+    fn new_internal(child: Root<K, V>, arena: &mut Arena<K, V>) -> Self {
+        let mut new_node = unsafe { InternalNode::new(arena) };
+        unsafe {
+            let new_node = arena.internal_node(new_node);
+            (*new_node).edges[0].write(child.node);
+        }
+        unsafe { NodeRef::from_new_internal(new_node, child.height + 1, arena) }
     }
 
     /// # Safety
     /// `height` must not be zero.
-    unsafe fn from_new_internal<A: Allocator + Clone>(
-        internal: Box<InternalNode<K, V>, A>,
+    unsafe fn from_new_internal(
+        internal: Id<InternalNode<K, V>>,
         height: usize,
+        arena: &Arena<K, V>,
     ) -> Self {
         debug_assert!(height > 0);
-        let node = NonNull::from(Box::leak(internal)).cast();
-        let mut this = NodeRef { height, node, _marker: PhantomData };
-        this.borrow_mut().correct_all_childrens_parent_links();
+        let mut this = NodeRef {
+            height,
+            node: internal,
+            _marker: PhantomData,
+        };
+        this.borrow_mut().correct_all_childrens_parent_links(arena);
         this
     }
 }
 
 impl<BorrowType, K, V> NodeRef<BorrowType, K, V, marker::Internal> {
     /// Unpack a node reference that was packed as `NodeRef::parent`.
-    fn from_internal(node: NonNull<InternalNode<K, V>>, height: usize) -> Self {
+    fn from_internal(node: Id<InternalNode<K, V>>, height: usize) -> Self {
         debug_assert!(height > 0);
-        NodeRef { height, node: node.cast(), _marker: PhantomData }
+        NodeRef {
+            height,
+            node,
+            _marker: PhantomData,
+        }
     }
 }
 
 impl<BorrowType, K, V> NodeRef<BorrowType, K, V, marker::Internal> {
     /// Exposes the data of an internal node.
     ///
     /// Returns a raw ptr to avoid invalidating other references to this node.
-    fn as_internal_ptr(this: &Self) -> *mut InternalNode<K, V> {
-        // SAFETY: the static node type is `Internal`.
-        this.node.as_ptr() as *mut InternalNode<K, V>
+    ///
+    /// The raw pointer does not keep a borrow on the arena, so extreme care
+    /// must be taken to avoid doing something that might cause the arena's
+    /// internal items to resize between getting this pointer and dereferencing
+    /// it!!!
+    fn as_internal_ptr(this: &Self, arena: &Arena<K, V>) -> *mut InternalNode<K, V> {
+        unsafe { arena.internal_node(this.node) as *mut _ }
     }
 }
 
 impl<'a, K, V> NodeRef<marker::Mut<'a>, K, V, marker::Internal> {
     /// Borrows exclusive access to the data of an internal node.
-    fn as_internal_mut(&mut self) -> &mut InternalNode<K, V> {
-        let ptr = Self::as_internal_ptr(self);
-        unsafe { &mut *ptr }
+    fn as_internal_mut(&mut self, arena: &Arena<K, V>) -> &mut InternalNode<K, V> {
```

Comment:
Since this is putting a safe wrapper around the raw-pointer-returning layer, should we use this opportunity to tie the lifetime of the returned borrow to the arena as well? In other words, should we do

```rust
fn as_internal_mut<'a>(&'a mut self, arena: &'a Arena<K, V>) -> &'a mut InternalNode<K, V> { ... }
```
?